feat: block producer redesign

Mirko-von-Leipzig opened the pull request:
This is still a WIP.

Notably, this only demonstrates the transaction pool implementation and even there skips some parts like handling notes and nullifiers.

Does not include batch producer nor block producer; as I think they should be rather straight-forward but depend on getting the pool interface correct.

We should also discuss how we want to integrate this work once its complete. The work itself is fairly non-trivial so it may be best to keep the existing design in-place until this new design has sufficient confidence. aka I'm not going to replace any existing code until we are very ready - probably split across several PRs?

This work is also missing revoking/dropping transactions e.g. due to recency conditions. This is possible to add, but since we don't actually have this feature yet I've left it out.

### Design overview & intent

The pool tracks inflight transactions, batches and blocks. The definition of inflight is extended slightly to include all data that we consider not stale. This includes some number of blocks which we know to already be complete and in the store. This is done to minimize race conditions for new transactions inputs from the store (by providing several block's of overlap/grace).

Transactions, batches and blocks are all handled similarly. They're kept in a pool and are only removed once they're considered stale. Dependency graphs are bi-directional i.e. they store both children and parents (except blocks which are always sequential so no need). Once items are removed due to staleness all references to them are removed. This means all items/graph edges should always be in the pool which we enforce using `.expect(..)` as this would indicate a rather serious bug.

Transaction and batch selection works by tracking what I called "roots". A node is considered a root if all of its ancestors have already been processed e.g. a transaction root's parents are all already part of some inflight batch, and a batch root's parents are all already part of some inflight block.

### Outstanding work

I anticipate the need for a lot of tests unless we can find some way of abstracting parts of this. Though I think in general graph book keeping is just painful so we just need to be thorough.

The rest of the state needs to be added, and the block and batch producers need at least a basic implementation so we can modify the pool api accordingly (missing outputs).

We need to pay particular attention to code that modifies the graphs; and should consider more expensive assertations, or runtime audits until we have performance concerns when running live?

Improve naming and comments which I'll get to.

We need to decide what to do in case of failure. Right now we don't really have any options except abort all transactions forming part of the failure?

Given that this is quite a lot of fairly intertwined code, I'm thinking maybe we should split this PR up more once tests are written? e.g. PR 1 includes just `add_transaction` and its tests and then we build from there.

bobbinth commented on crates/block-producer/src/tx_pool/mod.rs:
```diff
@@ -0,0 +1,514 @@
+
+use crate::store::TransactionInputs;
+
+#[derive(Debug, Clone, Copy, PartialEq, Eq, PartialOrd, Ord)]
+pub struct BatchId(u64);
```
For batch ID we can use a hash of all transaction IDs that go into the batch (using some stable order). This way, two batches with the same set of transactions would end up having the same ID.

Mirko-von-Leipzig replied:
This would be ideal but I have concerns over how expensive that hash would be? This has to be performed inline meaning it extends the transaction pool lock.

bobbinth replied:
I think this depends on which hash function we use:

- If we go with something like BLAKE3, computing a hash of 16 transaction IDs should take less than 1 microsecond.
- If we use our arithmetization-friendly hash (RPO) it would be around 50 microseconds (or around 25 microseconds if/when we migrate to RPX hash function).

We could probably start with BLAKE3 and switch to RPO/RPX only if needed, but I think either way it may be OK as my expectation was that batch selection will probably take on the order of 1 millisecond anyways.

Mirko-von-Leipzig replied:
I've left this as is still. I think we may want to separate the concept of batch ID and batch job ID for the situation where we drop a batch and re-create the exact same one. This would result in two "inflight" batches with the same ID.

Maybe this is fine; will have to think about it some more. For now both count and hashed are still around.

bobbinth replied:
> I think we may want to separate the concept of batch ID and batch job ID for the situation where we drop a batch and re-create the exact same one. This would result in two "inflight" batches with the same ID.

As far as I can tell, this shouldn't happen because we can't put the same transaction into two different batches at the same time (or at least we shouldn't). So, we could have two batches with the same ID - but it shouldn't be at the same time.

My preference is still to use a hash-based ID as I think it is conceptually a bit simpler and could help us catch some errors (i.e., if we do detect 2 batches with the same ID in flight, something probably went wrong).

Mirko-von-Leipzig replied:
An example:

Two inflight batches `N` and `N+1` where `N+1` depends on `N`. 

Batch `N` fails to prove and we inform the mempool of this. The mempool requeues all transactions from `N` and `N+1`.

Select two new batches and we get `N` and `N+1` again. We now have three batches inflight, with a duplicate `N+1`.

Note that the batch producer has no idea of any dependencies so it cannot automatically drop the original `N+1`.

bobbinth commented on crates/block-producer/src/tx_pool/mod.rs:
```diff
@@ -0,0 +1,514 @@
+}
+
+pub struct TransactionPool {
+    /// The latest inflight state of each account.
+    ///
+    /// Accounts without inflight transactions are not stored.
+    account_state: BTreeMap<AccountId, (Digest, TransactionId)>,
```
Should the value be a triplet `(Digest, TransactionId, Option<BlockNumber>)`? The block number would be used to track the block at the new state was inserted into the store. This way, we'd be able to easily remove all states for blocks which are too far in the past.

Also, it may make sense to create a newtype for this to expose an interface that is tailored for our needs.

Mirko-von-Leipzig replied:
Currently I use the transaction ID of the stale block to check this. This has the benefit that the account ID is part of the transaction data so we only have to check the specific state instead of iterating over all of it.

bobbinth commented on crates/block-producer/src/tx_pool/mod.rs:
```diff
@@ -0,0 +1,514 @@
+    /// The number of the next block we hand out.
+    next_block: BlockNumber,
+
+    /// All transactions currently inflight.
+    ///
+    /// This includes those not yet processed, those in batches and blocks after the stale block.
+    tx_pool: BTreeMap<TransactionId, InflightTransaction>,
+
+    /// Transactions ready to be included in a batch.
+    ///
+    /// aka transactions whose parent transactions are already included in batches.
+    tx_roots: BTreeSet<TransactionId>,
```
I would maybe encapsulate these in a separate struct - something like `InFlightTransactions`.

Mirko-von-Leipzig replied:
I've create graph types to encapsulate these.

bobbinth commented on crates/block-producer/src/tx_pool/mod.rs:
```diff
@@ -0,0 +1,514 @@
+    /// The next batches ID.
+    next_batch_id: BatchId,
+
+    batch_pool: BTreeMap<BatchId, InflightBatch>,
+
+    /// Batches which are ready to be included in a block.
+    ///
+    /// aka batches who's parent batches are already included in blocks.
+    batch_roots: BTreeSet<BatchId>,
```
Similar to the previous comment: maybe this deserves its own struct.

bobbinth commented on crates/block-producer/src/tx_pool/mod.rs:
```diff
@@ -0,0 +1,514 @@
+        // outputs.
+        let mut parents = BTreeSet::new();
+
+        // Merge inflight state with inputs.
+        //
+        // This gives us the latest applicable state for this transaction.
+        // TODO: notes and nullifiers.
+        if let Some((state, parent)) = self.account_state.get(&account_update.account_id()) {
+            parents.insert(*parent);
+            inputs.account_hash = Some(*state);
+        }
+
+        // Verify transaction input state.
+        // TODO: update notes and nullifiers.
+        if inputs.account_hash.unwrap_or_default() != account_update.init_state_hash() {
+            return Err(AddTransactionError::InvalidAccountState {
+                current: inputs.account_hash.unwrap_or_default(),
+                expected: account_update.init_state_hash(),
+            });
+        }
+
+        // Transaction is valid, update inflight state.
+        // TODO: update notes and nullifiers.
+        let tx_id = transaction.id();
+        self.account_state
+            .insert(transaction.account_id(), (account_update.final_state_hash(), tx_id));
```
If `account_states` was a bespoke struct, we could move this logic there as something like:
```rust
impl AccountStates {
    pub fn add_update(
        &mut self,
        update: &AccountUpdate,
        tx_id: TransactionId,
        store_state: Option<Digest>,
    ) -> Result<Vec<TransactionId>, AddTransactionError> {

    }
}
```

bobbinth commented on crates/block-producer/src/tx_pool/mod.rs:
```diff
@@ -0,0 +1,514 @@
+        Ok(())
+    }
+
+    /// Returns at most `count` transactions and a batch ID.
+    ///
+    /// Transactions are returned in a valid execution ordering.
+    ///
+    /// Returns `None` if no transactions are available.
+    pub fn select_batch(&mut self, count: usize) -> Option<(BatchId, Vec<Arc<ProvenTransaction>>)> {
```
Why do we want to provide `count` as a parameter? I would have thought that it would be up to the logic in this method to figure out how many transactions to return (and that would be based on some config parameters of the transaction pool).

Also, maybe for the return type we should have a dedicated struct? Something like `SelectedBatch` or `ProposedBatch`.

Mirko-von-Leipzig replied:
I was unsure where `count` should live; could be configured as a part of the pool or as a value in the batch producer. Logistically it made sense to me that it might be a constraint in the batch producer. But one could also argue it might form part of the transaction selection strategy.

bobbinth commented on crates/block-producer/src/tx_pool/mod.rs:
We may want to track how many times a transaction made it into a batch (e.g., as `InQueue(usize)`). This way, if let's say, the transaction was batched 3 times but failed to make it into a block, we'd automatically drop it.

bobbinth reviewed (commented):
Looks good! Thank you! I didn't not review all the logic, but did go through the most of the high-level structure and left some small comments inline.

The main comment is that I'd probably prefer to have dedicated structs which encapsulate specific pieces of functionality, and then use these structs in the `TransactionPool`.

Mirko-von-Leipzig commented:
I've refactored quite a bit; and I've added implementations for the rpc `add_transaction` and a batch producer with a worker pool.

There is still quite a bit missing. Notably tests, checking all inputs, and the block producer implementation.

bobbinth commented on crates/block-producer/src/mempool/mod.rs:
```diff
@@ -0,0 +1,269 @@
+    /// The next batches ID.
+    next_batch_id: BatchJobId,
+
+    /// Blocks which are inflight or completed but not yet considered stale.
+    block_pool: BTreeMap<BlockNumber, Vec<BatchJobId>>,
```
This is for keeping track of recent blocks, right? If so, I'm wondering if we actually need this. Specifically, there are 3 things from recent blocks that we need to keep track of:

- Account states - basically `(AccountId, Digest)` tuples.
- Newly created notes - `NoteId`s, maybe with inclusion paths.
- Newly created nullifiers - these would be just `Nullifier` structs stored in a set or something like that.

As you work though adding note/nullifier tracking to the mempool, it might make sense to store them together and get rid of this `block_pool` field.

Mirko-von-Leipzig replied:
These track all inflight blocks as well as recently applied blocks. The only reason we need the recent blocks is to efficiently trim the current state once we consider them beyond the "stale" threshold of incoming transactions. Otherwise the in-memory state just grows indefinitely.

This would also be used if we have multiple inflight blocks and one fails. So I'm not sure if we do ever want to remove this? Unless we enforce that there is at most exactly one block inflight.

bobbinth replied:
> This would also be used if we have multiple inflight blocks and one fails. So I'm not sure if we do ever want to remove this? Unless we enforce that there is at most exactly one block inflight.

As mentioned in another comment, I would not allow multiple in-flight blocks for now.

> The only reason we need the recent blocks is to efficiently trim the current state once we consider them beyond the "stale" threshold of incoming transactions.

I think we should be able to do the same without tracking the actual blocks. For example, as soon as a block is committed, we'd remove all relevant data from transaction and batch graph (this shouldn't require examining the entire state). Then, we'd update the "in-flight" data structure to add block number of the recently committed block to relevant pieces of data (account updates, nullifiers, notes). At this time, we'd also remove the stale data associated with some old block.

The above really depends on how we define the structure for the in-flight data though.

Mirko-von-Leipzig replied:
Resolving this thread in favor of this related one [here](https://github.com/0xPolygonMiden/miden-node/pull/502#discussion_r1783830150) to keep discussing the state tracking options.

bobbinth commented on crates/block-producer/src/mempool/batch_graph.rs:
```diff
@@ -0,0 +1,179 @@
+}
+
+#[derive(Clone, Debug)]
+struct Node {
+    status: Status,
+    transactions: Vec<TransactionId>,
+    parents: BTreeSet<BatchJobId>,
+    children: BTreeSet<BatchJobId>,
+}
```
I will need to have 2 different variants of batches: `SelectedBatch` and `ProvenBatch`. The main difference between the two is that `ProvenBatch` would also eventually have a proof of correct construction attached to it (and there may be other differences as well). But I think we should make provisions for them now (even if the underlying structs are identical for now).

Maybe we can track the difference through the `Status` field - but we could have a proper enum too. Maybe something like:
```rust
struct Node {
    data: Batch,
    parents: BTreeSet<BatchJobId>,
    children: BTreeSet<BatchJobId>,
}

enum Batch {
    InFlight(SelectedBatch),
    Proven(ProventBatch),
    InBlock(ProvenBatch),
}
```

bobbinth commented on crates/block-producer/src/mempool/batch_graph.rs:
```diff
@@ -0,0 +1,179 @@
+        removed
+    }
+
+    /// Removes a set of batches from the graph without removing any descendents.
+    ///
+    /// This is intended to cull completed batches from stale blocks.
+    pub fn remove_stale(&mut self, batches: Vec<BatchJobId>) -> Vec<TransactionId> {
```
Similar comment as the one for transaction graph: maybe this could be `remove_committed()`.

bobbinth reviewed (commented):
Looks good! Thank you! This is not a full review, but I did leave some comments inline - mostly minor though. I think the biggest remaining thing is to incorporate note/nullifier tracking as this may affect the structure a bit.

Mirko-von-Leipzig commented on crates/block-producer/src/mempool/mod.rs:
```diff
@@ -0,0 +1,269 @@
+    /// Transactions are returned in a valid execution ordering.
+    ///
+    /// Returns `None` if no transactions are available.
+    pub fn select_batch(
+        &mut self,
+        count: usize,
+    ) -> Option<(BatchJobId, Vec<Arc<ProvenTransaction>>)> {
```
I think you can argue for both. The batch producer knows best about the performance capacity of its workers for example. Or how much load they're under; or whether some have gone offline or are unstable.

We can also consider `count` an upper bound or hint; mempool can still figure out an optimal number instead.

Given that we have none of the above requirements yet; I'll move it to the mempool. I am a bit concerned that its doing pretty much everything though 😅

bobbinth replied:
> I think you can argue for both. The batch producer knows best about the performance capacity of its workers for example. Or how much load they're under; or whether some have gone offline or are unstable.
> 
> We can also consider `count` an upper bound or hint; mempool can still figure out an optimal number instead.

Agreed - but I think when we get to this level, the info will need to be a bit more detailed (e.g., try to give me a batch which is heavy on consumed notes, but light on created notes). And it is probably better to define the hint parameter once we actually know what we are going to use.

Mirko-von-Leipzig commented on crates/block-producer/src/mempool/transaction_graph.rs:
```diff
@@ -0,0 +1,123 @@
+        self.try_make_root(id);
+    }
+
+    pub fn pop_for_batching(
+        &mut self,
+    ) -> Option<(Arc<ProvenTransaction>, BTreeSet<TransactionId>)> {
```
Yes this is just a trivial implementation for now. Whatever we decide to go with in the future will change this and probably add some additional metadata to the transactions/batches.

Mirko-von-Leipzig commented on crates/block-producer/src/mempool/transaction_graph.rs:
```diff
@@ -0,0 +1,123 @@
+        }
+    }
+
+    pub fn remove_stale(&mut self, transactions: Vec<TransactionId>) {
```
The transaction data could be removed as soon as the block is finalized. This does mean we need to make changes when new transactions enter as the "parent" transaction might no longer exist.

This is fine though it does remove an opportunity for verifying the graphs correctness at runtime since we can no longer distinguish between a broken graph and parent data just being removed. The benefit is trimming the memory usage earlier.

For this initial implementation my thought process was to ensure that the graph was always correct i.e. all indices/edges always fully exist until the entire block is removed once its stale.

bobbinth replied:
> The transaction data could be removed as soon as the block is finalized. This does mean we need to make changes when new transactions enter as the "parent" transaction might no longer exist.

Makes sense - though I was thinking we'd make this check before adding the transaction to the graph (i.e., we call `add_transaction()` only with a valid set of parents).

Mirko-von-Leipzig replied:
I'm not sure if we're talking past each other here; so to clarify:

Inflight state contains the state of all in-progress transactions as well as some number of recently committed blocks. The latter is required so that new transactions have some grace/overlap period to pre-fetch input data from the store.

State also contains it's source transaction ID so that new transactions can be added as children to form the dependency graph.

State from stale blocks must be pruned otherwise the in-memory state will grow indefinitely and essentially be a copy of all the state ever. So if block `N` becomes "stale" (the wording here is bad since its referring to the transaction inputs being considered stale but... 🤷), all state still attached to block `N` should be removed.

Transactions in committed blocks can be deleted immediately, however this leaves the state transaction ID dangling. Here are the options to handle this:

1. Only delete transaction data once the block is stale (current approach). New transactions can depend on transactions already in committed blocks which is a bit of a waste ito compute/lookups, but shouldn't be much. Highest memory usage since txn data hangs around the longest. But graph edges remain consistent - all edges will always be valid.
2. Delete data as soon as it is committed to store. This makes invalid tx parent edges possible everywhere, meaning we cannot use `tx_graph.get(parent).expect(..)` to assert graph conditions at runtime. Instead we must allow these to fail as allowed.
3. Delete data as soon as it is committed to store. Remove edges everywhere, except state. Only the state edges are now allowed to fail. The reason we do the latter is because the state data must remain active until the stale block is removed.
    1. We can amend this by having the state contain an optional txn ID which can be removed once the block is committed. However this means each block will touch the state twice. Once when it is committed, and again when it is stale.

Maybe as an analogy, what we would be doing is changing from multiple `Rc<Transaction>` to a single `Rc` with parent and state pointers becoming `rc::Weak` pointers. Which is fine; but we do lose the ability to sanity check at runtime with possible memory leaks, and edges that should exist but no longer do.

Mirko-von-Leipzig replied:
> (i.e., we call add_transaction() only with a valid set of parents).

I don't quite follow this point

bobbinth replied:
I was thinking of something along the lines of the 3rd option: once a block is committed we prune all transactions of this block from `tx_graph` and all batches of this block from `batch_graph`. We do keep the information about the block's created notes/nullifiers and updated accounts separately (i think this is what you refer to as `state`). These are removed at a later point in time when the block becomes "stale".

But I haven't thought about it as much as you, so - may be missing some details about this approach.

> > (i.e., we call add_transaction() only with a valid set of parents).
> 
> I don't quite follow this point

The idea here was that when we call `tx_graph.add_transaction()` the parents will contain only transactions which are currently in `tx_graph`.

Mirko-von-Leipzig replied:
> The idea here was that when we call tx_graph.add_transaction() the parents will contain only transactions which are currently in tx_graph.

Got it; yes we can do that.

> I was thinking of something along the lines of the 3rd option: once a block is committed we prune all transactions of this block from tx_graph and all batches of this block from batch_graph. We do keep the information about the block's created notes/nullifiers and updated accounts separately (i think this is what you refer to as state). These are removed at a later point in time when the block becomes "stale".
>
> But I haven't thought about it as much as you, so - may be missing some details about this approach.

I think superficially I considered the transaction-only approach as simpler, as it doesn't have to also handle blocks. 

As a mockup:

```rust
/// Represents state tracking for a single account using tx only.
struct AccountState {
    /// Transactions and states from all inflight txns and non-stale blocks.
    /// Required to allow reverting transactions.
    states: VecDequeu<(Digest, TransactionId)>,
}

impl AccountState {
    /// Rewinds the state of all the given transactions i.e. pop_back.
    /// Could also simplify and just make it a # of txns to revert which means
    /// less checking required.
    fn revert(&mut self, txs: BTreeSet<TransactionId>) { ... }

    /// Removes the states from the from the front of the queue i.e. pop_front.
    /// Could also simplify to # here.
    fn remove_stale(&mut self, txs: BTreeSet<TransactionId>) { ... }

    /// The current state of the account.
    fn current(&self) -> Option<&(Digest, TransactionId)> { self.states.last() }

    fn insert(&mut self, state: Digest, tx: TransactionId) { self.states.push_back((state, tx)); }
}
```

Versus with block support:

```rust
struct AccountState {
    inflight: VecDequeu<(Digest, TransactionId)>,
    /// Latest complete but not-stale block.
    block: Option<Digest>,
}
```

impl AccountState {
    fn revert(&mut self, txs: BTreeSet<TransactionId>) { ... }

    fn remove_stale(&mut self, state: Digest) { 
        // Remove self.block if its state matches the input.
    }

    fn current(&self) -> Option<&(Digest, Option<TransactionId>)> { 
        self.states.last()
            .map(..)                 // map tx -> Some
            .or(
                 self.block.map(...) // add tx -> None
            )
    }

    /// Using count prevents safety checks but does make it easier to implement.
    fn block_complete(&mut self, tx_count: usize) {
         self.block = self.states
             .drain(0..tx_count)
             .last();
    }
}
```

I was expecting the block handling to be more complex tbh. The nice thing about the latter is that edges are always valid. I'll try it out and see how it goes.

Mirko-von-Leipzig commented on crates/block-producer/src/mempool/mod.rs:
```diff
@@ -0,0 +1,269 @@
+        self.transactions.remove_stale(stale_transations);
+    }
+
+    pub fn block_failed(&mut self, block: BlockNumber) {
+        // TBD.. not quite sure what to do here yet. Presumably the caller has already retried this
+        // block so the block is just inherently broken.
+        //
+        // Given lack of information at this stage we should probably just abort the node?
+        // In the future we might improve the situation with more fine-grained failure reasons.
+    }
```
Right now I don't we can actually recover from this though. Current options as I see them:

1. Panic
2. Requeue the transactions and hope for a different set of batches (we could introduce randomness to help here)
3. Requeue batches 
4. Drop all transactions from the failed block

bobbinth replied:
I think in the short term, option 4 seems like a good one. This would allow dropping the "poisoned" transaction which makes the block production fail.

In the longer term, I think option 2 would be better and here we can keep a count of how many times a transaction has been re-queued. If it was re-queued more times than some limit (e.g., 3), we would drop it then. The idea is to reduce the number of transactions that we actually drop as presumably not all transactions in the failed block were bad.

Mirko-von-Leipzig replied:
Makes sense. Ideally for (2) we would have more fine grained failure conditions which might pinpoint the problematic transaction(s).

bobbinth commented on crates/block-producer/src/mempool/mod.rs:
```diff
@@ -0,0 +1,269 @@
+    /// Blocks which are inflight or completed but not yet considered stale.
+    block_pool: BTreeMap<BlockNumber, Vec<BatchJobId>>,
+
+    /// The current block height of the chain.
+    completed_blocks: BlockNumber,
```
> For example, we might call `select_block` to start proving the next block even before submitting the current block to the store. This would result in effectively two inflight blocks from the perspective of the mempool.

I would probably disallow this for now as we can't do this anyway (i.e., starting construction of the next block requires that the previous block has already been computed). I think the assumption we should make is that we don't start construction of the next block until the currently constructed block is committed.

Mirko-von-Leipzig replied:
> I would probably disallow this for now as we can't do this anyway (i.e., starting construction of the next block requires that the previous block has already been computed).

I do mean after the computation has completed, but before the block has been sent to the store. This would elide the store time.

1. select block `N` from mempool
2. compute & prove block `N`
3. select block `N+1` and begin proving
4. submit block `N` to store
5. notify mempool `N` is complete once store completes

But I'll keep it simpler for now; we must just be aware of this in the block producer then.

bobbinth replied:
Yeah - for now let's keep it simple:

1. Select block `N` from mempool.
2. Compute and prove block `N`.
3. Submit block `N` to the store.
4. Notify mempool `N` is complete once store completes.
5. Select block `N+1` from mempool
6. etc.

Mirko-von-Leipzig commented on crates/block-producer/src/mempool/inflight_state.rs:
```diff
@@ -0,0 +1,314 @@
+    /// Ordering is chronological from front (oldest) to back (newest).
+    inflight: VecDeque<(Digest, TransactionId)>,
+
+    /// The latest committed state.
+    ///
+    /// Only valid if the committed count is greater than zero.
+    committed_state: Digest,
+
+    /// The number of committed transactions.
+    ///
+    /// If this is zero then the committed state is meaningless.
+    committed_count: usize,
```
Alternatively we can just move this back into a `VecDequeue`.

Mirko-von-Leipzig commented on crates/block-producer/src/mempool/inflight_state.rs:
```diff
@@ -0,0 +1,314 @@
+///
+/// TODO: this is a terrible name.
+pub struct StateDiff {
+    /// The number of transactions that affected each account.
+    account_transactions: BTreeMap<AccountId, usize>,
```
We could do additional checks here if we stored the actual state transitions.

Mirko-von-Leipzig commented on crates/block-producer/src/mempool/mod.rs:
```diff
@@ -0,0 +1,275 @@
+    /// Transactions are returned in a valid execution ordering.
+    ///
+    /// Returns `None` if no transactions are available.
+    pub fn select_batch(&mut self) -> Option<(BatchJobId, Vec<TransactionId>)> {
```
This intentionally returns IDs for now.

Mirko-von-Leipzig commented on crates/block-producer/src/mempool/mod.rs:
```diff
@@ -0,0 +1,275 @@
+    /// # Panics
+    ///
+    /// Panics if there is already a block in flight.
+    pub fn select_block(&mut self) -> (BlockNumber, BTreeSet<BatchJobId>) {
```
Intentionally returns IDs (for now).

Mirko-von-Leipzig commented:
As discussed offline, this work will be done on feature branch `next-block-producer` until we can finalize the entire body of work.

I'll summarize what this PR contains, and what it purposefully punts for future PRs. Some of this future work results in less-than-stellar current naming/structuring as its going to change soon. I'll attempt to point these sore spots out in a self-review so we don't focus on them too hard.

### Notable changes from last review

- State tracking moved to its own module/types.
- Added "state diffs" as a way to commit/revert batches of transactions, instead of keeping transactions around until they're pruned.
- Block and batch builder examples.
- Independent batches in a block.
- Nullifier support.

### Included in this PR

- [x] dependency tracking at transaction, batch and block levels
- [x] basic batch selection
- [x] basic block selection (ensures batches within a block are independent)
- [x] tracking and reverting inflight state (account state & nullifiers)

### Excluded

- [ ] input and output note state tracking
- [ ] data propagation beyond the mempool
- [ ] batching, a simple worker pool is demonstrated on dummy data
- [ ] block production, a simple worker loop is demonstrated on dummy data
- [ ] tests
- [ ] replacing existing framework
- [ ] reverting individual transactions (should be simple)

### What's next

Adding stronger typing to transactions, notes and transaction inputs will inform the structure of things; and I believe note tracking will fall out of this naturally.

Following that, implementing the types for batches and blocks makes sense I think.

I also want to raise `info!` and `panic!` out of the `mempool`, and instead have those handled in the callers. This won't be trivially possible everywhere but I think it makes more sense that way.

Mirko-von-Leipzig commented on crates/block-producer/src/mempool/mod.rs:
```diff
@@ -0,0 +1,275 @@
+}
+
+#[derive(Debug, Clone, Copy, PartialEq, Eq, PartialOrd, Ord)]
+pub struct BlockNumber(u32);
```
This probably belongs in `base`? Something else I've been itching to do is new-type all of the different fields in the block header for example. Having all of them be `Digest`/`u32` isn't great.

bobbinth replied:
Yes, I think moving this to `miden-base` would be a good idea. Would you like to make a PR there?

For the other fields - also agreed, but maybe we do these incrementally over time?

Mirko-von-Leipzig replied:
Will do. And incrementally is good; no rush :) One can also over do it.

bobbinth commented on crates/block-producer/src/block_builder/mod.rs:
```diff
@@ -143,3 +145,33 @@ where
+    }
+
+    async fn build_and_commit_block(&self, batches: BTreeSet<BatchJobId>) -> Result<(), ()> {
+        todo!("Aggregate, prove and commit block");
```
This will contain building the block and saving it to the store, right?

For block building, since we don't actually prove anything yet, I am thinking we could put in an artificial delay of 3 - 5 seconds to get the conditions close to the real thing.

Mirko-von-Leipzig replied:
Correct. I was also considering injecting failures randomly as mentioned [here](https://github.com/0xPolygonMiden/miden-node/pull/502#discussion_r1789644365).

bobbinth commented on crates/block-producer/src/batch_builder/mod.rs:
```diff
@@ -206,3 +213,84 @@ where
+
+    fn spawn(&mut self, id: BatchJobId, transactions: Vec<TransactionId>) {
+        self.0.spawn(async move {
+            todo!("Do actual work like aggregating transaction data");
```
Similar to the above comment: batch building could also take 3 - 5 seconds. Let's artificially simulate this for now.

Mirko-von-Leipzig replied:
I was wondering if we should also simulate failures. So something like sleep a random period, and then randomly fail a batch every now and then.

If we do want this, the failure rate should probably be a configurable parameter. Similar for the entire block.

Though this might annoy users if this causes a user tx to expire. And could look bad ito throughput; though I don't think that's an issue right now (?).

bobbinth replied:
If this is configurable and not too difficult to add, I don't mind doing it.

bobbinth reviewed (approved):
Looks good! Thank you! Not a very thorough review from me - but I did leave a few comments inline (mostly for the future).

Mirko-von-Leipzig commented on crates/block-producer/src/server/mod.rs:
```diff
@@ -105,3 +122,114 @@ impl BlockProducer {
+        );
+        debug!(target: COMPONENT, proof = ?tx.proof());
+
+        // TODO: These steps should actually change the type of the transaction.
+        // TODO: How expensive is this step? Should it be spawn_blocking?
+        TransactionVerifier::new(MIN_PROOF_SECURITY_LEVEL).verify(tx.clone())?;
```
I was unsure if we can always implicitly trust the RPC component.

I'll remove the verification then.
